### PR TITLE
Bugfix/corp updates

### LINKED ISSUE #3003: Bug: (Bitnode 8) Insane Corporation payout with dividend of 1e-99
In bitnode 8 i founded a corp to see what happens.
Unexpectedly there were payouts (but very very bad).
Surprisingly payouts decreased with increasing dividend percentage.
So I tried what happens when I go below 1 percent (0,000000000000001 in german locale) and, as assumed, payouts increased.
When I committed that dividend and reopened the dividend window the fraction of zero was presented in exponetial form (1e-15 or similar).
Then I modified the dividend to 1e-99 and got a payout of several $100t in a company with a revenue of under $1m/s.

Savegame: [bitburnerSave_1645208094_BN8x0.zip](https://github.com/danielyxie/bitburner/files/8099482/bitburnerSave_1645208094_BN8x0.zip)
Version: v1.4.0 49de4a28 (steam version)

### LINKED ISSUE #3007: Corrupted save after entering a (wrong?) price formula in corporation mode
### How did this happen?
After setting a sell price for a software  product (Software division in corporation mode) the game went into recovery mode.
Any further attempt to acces this corp division result in the game going into recovery mode.
I might have written an invalid character or formula in the product price field.

Please fill this information with details if relevant.

- [x] Save file
- [x] Steps to reproduce: load the attached save, go to Corp tab, open the "Software div" division of the corporation.

### Environment

* Error: SyntaxError: Unexpected number
* Page: Corporation
* Version: v1.4.0 (49de4a28)
* Environment: Production
* Platform: Steam
* UserAgent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) bitburner/1.0.0 Chrome/93.0.4577.82 Electron/14.0.2 Safari/537.36
* Features: lang=en-US cookiesEnabled=true doNotTrack=null indexedDb=true
* Source: n/a

### Save
[bitburnerSave_1645393681_BN3x0.json.txt](https://github.com/danielyxie/bitburner/files/8104979/bitburnerSave_1645393681_BN3x0.json.txt)

PS: I'd like to recover my save if possible... Would gladly do it myself if i'd knew what the save file content would mean. If not I have a backup before entering the bitnode, it's a couple week of playtime but I can live with it.

### LINKED ISSUE #3173: Bulk Purchase errors when not affordable, but code can't see price
[Bulk Purchase](https://github.com/danielyxie/bitburner/blob/fc963bb583f31c053a26074f05e29dc1ad5de905/src/Corporation/Actions.ts#L265)
How is a player's code supposed to avoid triggering the error that is thrown when the corporation can't afford a bulk purchase, if the code can't see what the price is?
There seems to be no limit to how much the price can change, so a reasonable hardcoded estimate will eventually become wrong.
And calculating the price of a material that is not used for production appears to be very impractical if at all possible for player code currently.

### LINKED ISSUE #3174: There is no netscript function to limit corporation production
https://github.com/danielyxie/bitburner/blob/fc963bb583f31c053a26074f05e29dc1ad5de905/src/Corporation/Actions.ts#L479

There is currently no netscript function to set the production limit of corporation products (despite there being a ui function for doing this manually), which becomes an issue when there is not enough space in the warehouse to support the full production of the city's products, and not enough funds to sufficiently expand the warehouse. This may just be a late game problem, but still.

Is this intentionally not available to scripts because it could enable price inflation to get more extreme than it already does?

### LINKED ISSUE #3175: Leaving production limit box empty does not set it to no limit
https://github.com/danielyxie/bitburner/blob/fc963bb583f31c053a26074f05e29dc1ad5de905/src/Corporation/ui/LimitProductProductionModal.tsx#L32

Regarding how the limit production modal for a corporation product currently works and does not work as stated:

Since the onChange event only fires when the text in the box changes, leaving the box empty on a freshly opened modal and then pressing enter or clicking the button does not change the limit to no limit if there was a limit set previously.

Currently the only ways to change the limit to no limit is to either put a negative number in the box, or typing something into the box and then deleting it again before pressing enter or clicking the button.

### LINKED ISSUE #3180: Producing too many materials lowers product production to 0
https://github.com/danielyxie/bitburner/blob/77cb7dd775b0d738bd1c945cdcf4b1e2836301e8/src/Corporation/Industry.ts#L449

With apparently too high production multiplier, because materials are processed before products, the produced products may fill the entire warehouse, causing there to be 0 production of products. This has only been observed to occur in the Real Estate industry, especially with production multiplier over 100k.
This causes the profits of the division to be much lower than they should be.
Since there is currently no way to artificially limit production of materials like there is for products, the only way currently to prevent this from happening is to lower the amount of production boosting materials in the warehouse until material production is low enough to not take up all the available space.

If possible, determining the produced amount of materials and products should be done together, both so that the production equation becomes factual as shown, with equal amounts produced between materials and each product (when not set to limit production), and most importantly so that material production can not drown out all product production.
Ideally this would also let there be equal production amounts (when limiting production is not set) between the products of the industry instead of the current first come first serve basis where the oldest products are made first, and newer ones only made if there is space left over.

### LINKED ISSUE #3205: dev menu: add button to quickly create gang / corp / join blade

### LINKED ISSUE #3264: "limit production" on corporate products can not be undone
despite the text says that leaving the box blank, it is not possible to unset the production limit.

workaround is to set it to high value

Bitburner 1.5.0 aabb764d

### LINKED ISSUE #3288: Can bulk purchase too much via ui?
image

if i was to guess maybe it hit NAN

### LINKED ISSUE #3312: corporation.expandCity() resets Office values if expansion already exists
The function corporation.expandCity() resets the office of a divison expansion in a city, when it is called and the expansion already exists.

The amount of workers is set back to zero and the maximum worker amount is set back to 3. Warehouse and other stuff remains untouched.

Steps to reproduce:
-> Divison office in city is upgraded to 6/6 workers
-> run function like "corporation.expandCity(divName, city)"
-> Office of divison in city is 0/3 workers

Version: v1.6.0 (a0361c9a)

image
image

### LINKED ISSUE #3334: Cannot bulk purchase
I always get the message "Error: You do not have enough warehouse size to fit this purchase"

If i'd have to guess, then i'd guess it's because this: https://github.com/danielyxie/bitburner/blob/dev/src/Corporation/Actions.ts#L261
should be `amt * matSize > maxAmount` instead of `amt * matSize <= maxAmount`.

### LINKED ISSUE #3359: [Feature Request] Corporations - Add a cancel option to in-progress products in development
Sometimes one accidentally enters the values incorrectly into the product development budget/advertisement budget and hits confirm too fast. Then you have to wait the entire development cycle to redo it, if your list of products is full.

It's been pointed out to me that this is unlockable with the research "uPgrade: Dashboard". As such, wellwhat and I agreed that perhaps it would be more appropriate, as it's potentially a little frustrating to have the cancel part locked behind an upgrade, to instead:

Allow the cancel without any researches, but you lose any funds invested in the development of the product. Once "uPgrade: Dashboard" is unlocked, you then get some percentage (or all) of the funding back when you discontinue during development.

### PR DESCRIPTION
**Pull request by phyzical**
closes #3174
closes #3312 
closes #3205
closes #3007
closes #3003 
closes #3180
closes #3173
closes #3175
closes #3334
closes #3288
closes #3264
closes #3359

* added check for already expanding into city
image

* add logic in dev menu to support easy corp, gang and bladeburner
Apr-02-2022 13-00-58
* removes spaces in at start and end of product name on creation
Apr-02-2022 13-02-52
* dissalow corporations when multiplier is below 0.15
Apr-02-2022 13-04-02
* move modals to a modals folder
* added logic to support material production limit
Apr-02-2022 13-07-00
* expose material cost
image
* added the ability to mass upgrade warehouse
Apr-02-2022 13-37-14
* fix for reseting limit production
Apr-02-2022 13-33-59

### CODE REVIEW

**Comment by TheRealMaxion** [comment]
Also closes #3264

**Comment by hydroflame** [comment]
PR made before mega fmt

run `npm run format` and push to (probably) fix merge conflicts.

File: src/Corporation/Actions.ts [hunk 1]
```diff
@@ -345,10 +347,17 @@ export function PurchaseWarehouse(corp: ICorporation, division: IIndustry, city:
...
 }
 
-export function UpgradeWarehouse(corp: ICorporation, division: IIndustry, warehouse: Warehouse): void {
-  const sizeUpgradeCost = CorporationConstants.WarehouseUpgradeBaseCost * Math.pow(1.07, warehouse.level + 1);
+export function UpgradeWarehouseCost(warehouse: Warehouse, amt: number): number {
+  return Array.from(Array(amt).keys()).reduce(
+    (acc, index) => acc + CorporationConstants.WarehouseUpgradeBaseCost * Math.pow(1.07, warehouse.level + 1 + index),
+    0,
+  );
+}
```
Comment (Ornedan) [inline src/Corporation/Actions.ts:355]:
```js
let cost = 0;
for(let n = 1; n <= amt; n++)
  cost += CorporationConstants.WarehouseUpgradeBaseCost * Math.pow(1.07, warehouse.level + n);
return cost;
```
Would IMO be more readable.

Doing it with a functional approach would require a few extra basic functions in order to make the code straightforward:
`interval(1. amt).map(n => CorporationConstants.WarehouseUpgradeBaseCost * Math.pow(1.07, warehouse.level + n).sum()`

File: src/NetscriptFunctions/Corporation.ts [hunk 2]
```diff
@@ -311,12 +317,13 @@ export function NetscriptCorporation(
       checkAccess("getPurchaseWarehouseCost", 7);
       return CorporationConstants.WarehouseInitialCost;
     },
-    getUpgradeWarehouseCost: function (_divisionName: unknown, _cityName: unknown): number {
+    getUpgradeWarehouseCost: function (_divisionName: unknown, _cityName: unknown, _amt: unknown = 1): number {
       checkAccess("upgradeWarehouse", 7);
       const divisionName = helper.string("getUpgradeWarehouseCost", "divisionName", _divisionName);
       const cityName = helper.city("getUpgradeWarehouseCost", "cityName", _cityName);
+      const amt = helper.number("getUpgradeWarehouseCost", "amount", _amt);
```
Comment (Ornedan) [inline src/NetscriptFunctions/Corporation.ts:324]: Needs a `amt >= 1` sanity-check. Either here or in the action.

File: src/NetscriptFunctions/Corporation.ts [hunk 3]
```diff
@@ -376,12 +384,13 @@ export function NetscriptCorporation(
       const corporation = getCorporation();
       PurchaseWarehouse(corporation, getDivision(divisionName), cityName);
     },
-    upgradeWarehouse: function (_divisionName: unknown, _cityName: unknown): void {
+    upgradeWarehouse: function (_divisionName: unknown, _cityName: unknown, _amt: unknown): void {
       checkAccess("upgradeWarehouse", 7);
       const divisionName = helper.string("upgradeWarehouse", "divisionName", _divisionName);
       const cityName = helper.city("upgradeWarehouse", "cityName", _cityName);
+      const amt = helper.number("upgradeWarehouse", "amount", _amt);
```
Comment (Ornedan) [inline src/NetscriptFunctions/Corporation.ts:391]: Also needs the `amt => 1` sanity-check.

File: src/ScriptEditor/NetscriptDefinitions.d.ts [hunk 4]
```diff
@@ -6661,8 +6661,9 @@ export interface WarehouseAPI {
    * Upgrade warehouse
    * @param divisionName - Name of the division
    * @param cityName - Name of the city
+   * @param amt - amount of upgrades defaults to 1
    */
-  upgradeWarehouse(divisionName: string, cityName: string): void;
+  upgradeWarehouse(divisionName: string, cityName: string, amt: number): void;
```
Comment (Ornedan) [inline src/ScriptEditor/NetscriptDefinitions.d.ts:6666]: `amt` is being optional should be noted in the signature: `amt?: number`

File: src/ScriptEditor/NetscriptDefinitions.d.ts [hunk 5]
```diff
@@ -6678,16 +6679,35 @@ export interface WarehouseAPI {
...
   getPurchaseWarehouseCost(): number;
   /**
    * Gets the cost to upgrade a warehouse to the next level
+   * @param divisionName - Name of the division
+   * @param cityName - Name of the city
+   * @param amt - amount of upgrades defaults to 1
    * @returns cost to upgrade
    */
-  getUpgradeWarehouseCost(adivisionName: any, acityName: any): number;
+  getUpgradeWarehouseCost(adivisionName: any, acityName: any, amt: number): number;
```
Comment (Ornedan) [inline src/ScriptEditor/NetscriptDefinitions.d.ts:6710]: `amt` is being optional should be noted in the signature: `amt?: number`

File: src/NetscriptFunctions/Corporation.ts [hunk 6]
```diff
@@ -74,6 +78,8 @@ export function NetscriptCorporation(
     if (!player.canAccessCorporation() || player.hasCorporation()) return false;
     if (!corporationName) return false;
     if (player.bitNodeN !== 3 && !selfFund) throw new Error("cannot use seed funds outside of BitNode 3");
+    if (BitNodeMultipliers.CorporationSoftCap < 0.15)
+      throw new Error(`You cannot create a corporation for Bitnode ${player.bitNodeN}`);
```
Comment (Ornedan) [inline src/NetscriptFunctions/Corporation.ts:82]: "You cannot create a corporation **_in_** Bitnode"

**Pull request merged**